torchdata/datapipes/iter/util/prefetcher.py

```diff
@@ -138,3 +141,94 @@ def resume(self):
+            # The mapping type may not support `__init__(iterable)`.
+            return pinned_data
+    elif isinstance(data, tuple) and hasattr(data, "_fields"):  # namedtuple
+        return type(data)(*(pin_memory(sample, device) for sample in data))
+    elif isinstance(data, tuple):
+        return [pin_memory(sample, device) for sample in data]  # Backwards compatibility.
+    elif isinstance(data, collections.abc.Sequence):
+        pinned_data = [pin_memory(sample, device) for sample in data]
+        try:
+            return type(data)(pinned_data)
+        except TypeError:
+            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
+            return pinned_data
+    else:
+        return data
+
+
+@functional_datapipe("pin_memory")
+class PinMemoryIterDataPipe(IterDataPipe):
+    r""" """
+
+    def __init__(self, source_datapipe, pin_memory_fn=pin_memory_fn):
+        self.pin_memory_fn = pin_memory_fn
+        self.thread: Optional[threading.Thread] = None
+        self.prefetch_data: Optional[_PrefetchData] = None
+
+    def is_replicable(self) -> bool:
+        return False
+
+    @staticmethod
+    def thread_worker(prefetch_data: _PrefetchData):
+        itr = iter(prefetch_data.source_datapipe)
+        while not prefetch_data.stop_iteration:
+            while prefetch_data.run_prefetcher:
+                if len(prefetch_data.prefetch_buffer) < prefetch_data.buffer_size:
+                    try:
+                        data = self.pin_memory_fn(next(itr))
+                        prefetch_data.prefetch_buffer.append(data)
+                    except Exception as e:
+                        prefetch_data.prefetch_buffer.append(e)
+                        prefetch_data.run_prefetcher = False
+                else:  # Buffer is full, waiting for main thread to consume items
+                    # TODO: Calculate sleep interval based on previous consumption speed
+                    time.sleep(PRODUCER_SLEEP_INTERVAL)
+            time.sleep(PRODUCER_SLEEP_INTERVAL)
+
+    def __iter__(self):
+        try:
+            prefetch_data = _PrefetchData(self.source_datapipe, 1)
+            self.prefetch_data = prefetch_data
+            thread = threading.Thread(target=PrefetcherIterDataPipe.thread_worker, args=(prefetch_data,), daemon=True)
+            thread.start()
+            self.thread = thread
+
+            while not prefetch_data.stop_iteration:
+                if len(prefetch_data.prefetch_buffer) > 0:
+                    data = prefetch_data.prefetch_buffer.popleft()
+                    if isinstance(data, Exception):
+                        prefetch_data.run_prefetcher = False
+                        prefetch_data.stop_iteration = True
+                        if not isinstance(data, StopIteration):
+                            raise data
+                    else:
+                        yield data
+                else:
+                    time.sleep(CONSUMER_SLEEP_INTERVAL)
```

Comment:
For reference, what do you think about the control flow here? @NivekT

Reply:
It might not be totally correct. Let me take a deeper look

Reply:
I just noticed no pinning is happening yet. Is this meant to replace the existing control flow for `Prefetcher`? 

If so, see this [comment](https://github.com/pytorch/data/pull/879#discussion_r1107757219).

Reply:
Updated this PR to reflect what's in my mind.